### ISSUE TITLE
Maximum recursion depth when using ParamSpec/Concatenate

### ISSUE THREAD

**Issue by karlicoss** [python/mypy#15388]
**Crash Report**

I've been trying to annotate some decorators using `ParamSpec`/`Concatenate` and got mypy to segfault. In particular was trying to improve type annotations for https://github.com/python/typeshed/blob/main/stubs/decorator/decorator.pyi -- currently they are a bit too generic

**Traceback**

No traceback on latest stable mypy, just segfault
If I use `master` on mypy playground I get INTERNAL_ERROR and this:

```
  ...
  File "/layers/google.python.pip/pip/lib/python3.10/site-packages/mypy/expandtype.py", line 536, in expand_types
    a.append(t.accept(self))
  File "/layers/google.python.pip/pip/lib/python3.10/site-packages/mypy/types.py", line 1917, in accept
    return visitor.visit_callable_type(self)
  File "/layers/google.python.pip/pip/lib/python3.10/site-packages/mypy/expandtype.py", line 429, in visit_callable_type
    arg_types = self.expand_types(t.arg_types)
  File "/layers/google.python.pip/pip/lib/python3.10/site-packages/mypy/expandtype.py", line 536, in expand_types
    a.append(t.accept(self))
  File "/layers/google.python.pip/pip/lib/python3.10/site-packages/mypy/types.py", line 1917, in accept
    return visitor.visit_callable_type(self)
  File "/layers/google.python.pip/pip/lib/python3.10/site-packages/mypy/expandtype.py", line 382, in visit_callable_type
    param_spec = t.param_spec()
  File "/layers/google.python.pip/pip/lib/python3.10/site-packages/mypy/types.py", line 2035, in param_spec
    prefix = Parameters(self.arg_types[:-2], self.arg_kinds[:-2], self.arg_names[:-2])
  File "/layers/google.python.pip/pip/lib/python3.10/site-packages/mypy/types.py", line 1569, in __init__
    super().__init__(line, column)
  File "/layers/google.python.pip/pip/lib/python3.10/site-packages/mypy/types.py", line 225, in __init__
    super().__init__(line, column)
RecursionError: maximum recursion depth exceeded while calling a Python object
```

**To Reproduce**

Reproduces in https://mypy-play.net/?mypy=master&python=3.11&flags=show-traceback

```
from typing import ParamSpec, Concatenate, TypeVar, Callable


TD = TypeVar('TD')
PD = ParamSpec('PD')
def wrapper(
    caller: Callable[Concatenate[Callable[PD, TD], PD], TD]
) -> None:
    return None


T = TypeVar('T')
P = ParamSpec('P')
def _wrapped(func: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
    return func(*args, **kwargs)


_wrapped = wrapper(_wrapped)
```

**Your Environment**

Segfaults on latest stable mypy (`1.3.0`) regardless python version (tried 3.10/3.11/3.12)
`INTERNAL_ERROR` on `master` mypy

**Comment by AlexWaygood** [comment]
Haven't looked in depth, but this is probably the same thing as #15073 (cc. @A5rocks)

**Comment by A5rocks** [comment]
The crash specifically also gets fixed by https://github.com/python/mypy/pull/15272 (yay 🎉) but the example still has some incorrect (?) behavior that I haven't thought about:
```
@A5rocks ➜ /workspaces/mypy (fix-paramspec-constraints) $ mypy repro.py
repro.py:18: error: "wrapper" does not return a value  [func-returns-value]
Found 1 error in 1 file (checked 1 source file)
```

Sticking a `reveal_type` in there makes the wonkiness even stranger:
```
@A5rocks ➜ /workspaces/mypy (fix-paramspec-constraints) $ mypy repro.py
repro.py:18: error: Incompatible types in assignment (expression has type "None", variable has type "Callable[[Callable[P, T], **P], T]")  [assignment]
repro.py:18: note: Revealed type is "None"
Found 1 error in 1 file (checked 1 source file)
```

tl;dr kinda fixed (papered over?) by https://github.com/python/mypy/pull/15272 so that's cool.
